lib/assert.js

```diff
@@ -341,6 +359,233 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
   }
 };
 
+function isPlainObject(obj) {
+  if (typeof obj !== 'object' || obj === null) {
+    return false;
+  }
+  const proto = ObjectGetPrototypeOf(obj);
+  return proto === ObjectPrototype || proto === null || ObjectPrototypeToString(obj) === '[object Object]';
+}
+
+function assertIncludes(actual, expected, loose) {
+  if (expected.length > actual.length) {
+    return false;
+  }
+  if (typeof actual === 'string') {
+    return StringPrototypeIncludes(actual, expected);
+  }
+
+  return ArrayPrototypeEvery(expected, (item) => ArrayPrototypeSome(actual, (actualItem) => {
+    // eslint-disable-next-line eqeqeq
+    return loose ? actualItem == item : actualItem === item;
+  }));
+}
+
+/**
+ * Compares two objects or values recursively to check if they are equal.
+ * @param {any} actual - The actual value to compare.
+ * @param {any} expected - The expected value to compare.
+ * @param {boolean} [loose=false] - Whether to use loose comparison (==) or strict comparison (===). Defaults to false.
+ * @param {Set} [comparedObjects=new Set()] - Set to track compared objects for handling circular references.
+ * @returns {boolean} - Returns `true` if the actual value matches the expected value, otherwise `false`.
+ * @example
+ * // Loose comparison (default)
+ * compareBranch({a: 1, b: 2, c: 3}, {a: 1, b: '2'}); // true
+ *
+ * // Strict comparison
+ * compareBranch({a: 1, b: 2, c: 3}, {a: 1, b: 2}, true); // true
+ */
+function compareBranch(
+  actual,
+  expected,
+  loose = false,
+  comparedObjects = new SafeSet(),
+) {
+  // Check for Map object equality
+  if (isMap(actual) && isMap(expected)) {
+    if (actual.size !== expected.size) {
+      return false;
+    }
+    const safeIterator = FunctionPrototypeCall(SafeMap.prototype[SymbolIterator], actual);
+    const safeIteratorArray = ArrayFrom(safeIterator);
+
+    for (let i = 0; i < safeIteratorArray.length; i++) {
+      const { 0: key, 1: val } = safeIteratorArray[i];
```

Comment:
nit: we don't need to create a throw-away array here
```suggestion

    for (const { 0: key, 1: val } of safeIterator) {
```

Reply:
@aduh95 this was intentionally done to comply with https://github.com/nodejs/node/blob/b0443690f89816ca3b9b8dee7e73f4527b3ff3e0/doc/contributing/primordials.md#unsafe-array-iteration

Reply:
As I said in https://github.com/nodejs/node/pull/54630#discussion_r1766362234, this doesn't apply here, and my suggestion still stands :)